File: 2416-sum-of-prefix-scores-of-strings/2416-sum-of-prefix-scores-of-strings.py
class Solution:
    def sumPrefixScores(self, words):
        n = len(words)
        sorted_indices = list(range(n))
        
        sorted_indices.sort(key=lambda x: words[x])
        
        common_prefix_lengths = self.calculateCommonPrefixLengths(words, sorted_indices)
        return self.calculateScores(words, sorted_indices, common_prefix_lengths)

    def calculateCommonPrefixLengths(self, words, sorted_indices):
        lengths = [0] * len(words)
        for i in range(1, len(words)):
            prev = words[sorted_indices[i - 1]]
            curr = words[sorted_indices[i]]
            common_length = 0
            while (common_length < len(prev) and 
                   common_length < len(curr) and 
                   prev[common_length] == curr[common_length]):
                common_length += 1
            lengths[i] = common_length
        return lengths

    def calculateScores(self, words, sorted_indices, common_prefix_lengths):
        scores = [0] * len(words)
        for i in range(len(sorted_indices)):
            idx = sorted_indices[i]
            length = len(words[idx])
            scores[idx] += length
            j = i + 1
            common_length = length
            while j < len(words):
                common_length = min(common_length, common_prefix_lengths[j])
                if common_length == 0:
                    break
                scores[idx] += common_length
                scores[sorted_indices[j]] += common_length
                j += 1
        return scores

File: 2416-sum-of-prefix-scores-of-strings/2416-sum-of-prefix-scores-of-strings.cpp
class Solution {
public:
    vector<int> sumPrefixScores(vector<string>& words) {
        int n = words.size();
        vector<int> sortedIndices(n);
        for (int i = 0; i < n; i++) sortedIndices[i] = i;

        sort(sortedIndices.begin(), sortedIndices.end(), [&words](int a, int b) {
            return words[a] < words[b];
        });

        vector<int> commonPrefixLengths = calculateCommonPrefixLengths(words, sortedIndices);
        return calculateScores(words, sortedIndices, commonPrefixLengths);
    }

private:
    vector<int> calculateCommonPrefixLengths(const vector<string>& words, const vector<int>& sortedIndices) {
        vector<int> lengths(words.size(), 0);
        for (int i = 1; i < words.size(); i++) {
            const auto& prev = words[sortedIndices[i - 1]], &curr = words[sortedIndices[i]];
            int commonLength = 0;
            while (commonLength < prev.length() && commonLength < curr.length() && prev[commonLength] == curr[commonLength]) {
                commonLength++;
            }
            lengths[i] = commonLength;
        }
        return lengths;
    }

    vector<int> calculateScores(const vector<string>& words, const vector<int>& sortedIndices, const vector<int>& commonPrefixLengths) {
        vector<int> scores(words.size(), 0);
        for (int i = 0; i < sortedIndices.size(); i++) {
            int idx = sortedIndices[i], len = words[idx].length();
            scores[idx] += len;
            int j = i + 1, commonLength = len;
            while (j < words.size()) {
                commonLength = std::min(commonLength, commonPrefixLengths[j]);
                if (commonLength == 0) break;
                scores[idx] += commonLength;
                scores[sortedIndices[j]] += commonLength;
                j++;
            }
        }
        return scores;
    }
};
